### ISSUE TITLE
Support for passive streaming of results instead of active writing by IActionResultExecutor

### ISSUE THREAD

**Issue by Danielku15** [aspnet/Mvc#8307]
#### Is this a Bug or Feature request?: 

Feature request

#### Description of the problem:

In the old ASP.net Web API the `IHttpActionResults` were baked by the classical `HttpResponseMessage` and `HttpContent`. The `HttpContent` provides the nice `ReadAsStreamAsync()` which allows the service stack (server) to decide how this data packed and submitted further. 

In ASP.net core and driven by OWIN this communication direction was inverted. The `IServer` needs to provide already a response `Stream` (owin.ResponseBody) where the `IActionResultExecutor` directly will write to via `context.Response.Body`. 

This works fine for most of the server implementations provided by the ASP.net infrastructure. But not all communication systems have a raw stream where you can directly write to. The data might need to be packed into an additional data frame. In the old ASP.net Web API you simply passed on the stream for the body to the component which was then encoding header and body appropriatly. 

ASP.net core should also provide a way that ActionExecutors do not directly write the whole body into a result stream but rather set a result-stream object like in the old ASP.net WebApi. (see proposal below) 

For most results it is no problem to buffer the whole result in-memory before transmitting it. But especially the File- and Stream results should provide the raw stream to read from to the server or middleware. 

##### Concrete usecase

We have the strict security requirement that the ASP.net Web API is running in the LAN, not accessible via internet. This way the service has access to internal databases and other servers/services. The OData based API should then be exposed via a reverse proxy / relay running in the DMZ. 

Our classical solution for this is to have a WCF Duplex channel where we connect from the LAN to the DMZ service. The DMZ service accepts the HTTP requests, and tunnels the request through the duplex channel to the backend service. There I start the Web API using a custom IServer implementation. Whenever a request is incoming via WCF I forward it to the Web API via OWIN. 

And here you can clearly see the issue. If you have a `FileStreamResult` or `ObjectResult` which is executed as part of the request, it will directly set the headers and write the whole body data into the body stream. But from a WCF perspective you do not have a stream you can directly write to. 

Once the whole request is completed, you pass on the result object to WCF to sent it as reply. 

In my case it is a WCF channel where the result object needs to be sent to but other servers could require: 
- A service bus for dispatching the results
- A websocket based channel (HTTP over WebSocket tunnel)

##### Implementation Idea
 
ASP.net core (Web API) could define and implement a OWIN extension that allows passing a HTTP body stream from the application to the OWIN environment. Based on the sendFile extension::

| Key Name | Description |
|-------------|--------------|
| sendfile.SendStreamAsync | The SendStreamFunc provided by the server or middleware for the application to utilize. See Delegate Signature |

**Delegate Signature**
`using SendStreamFunc = Func<Stream /* stream*/, Task /*complete*/>;`

**Consumption**

- The server or middleware provides the SendStreamFunc in the environment dictionary. 
- The consumer checks for its presence and invokes it once for the data to be send. 
- Before, after, or in between SendStream operations the consumer MUST NOT write to the response body stream. 
- Before, after, or in between SendStream operations the consumer MAY define additional response headers. 
- The server or middleware implementing the SendStreamFun MUST NOT complete the corresponding Task until it is completely finished accessing the stream. 
- The server or middleware implementing the SendStreamFun MUST NOT dispose the application context before the send operation is complete and the stream is still used. 
- As part of the task completion the consumer MUST take action to free any resources and destroy the stream. 

##### Current workaround 
The main solution currently is to rely on a buffered message transmission via MemoryStream in such servers. For physical file transmissions you can also implement the OWIN SendFile extension in your server/middleware. 

For raw streams (that might pull data from another source on-demand) no clean solution exists. Maybe complex deferred executions and active stream pipes could also achieve a active-to-passive conversion.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @Danielku15.
@Tratcher, what are you thoughts regarding this ask?

@Danielku15, if you're looking for just buffering, consider using the `ResponseBufferingMiddleware` from the https://www.nuget.org/packages/Microsoft.AspNetCore.Buffering/ package.

**Comment by Tratcher** [comment]
There are a few different ways you could achieve your goal. The simplest and most compatible would be for your custom IServer to implement a custom Stream rather than using MemoryStream. That's how all of the real servers work. The lower level APIs our servers work with rarely provide a Stream implementation themselves, the server provides one on top of some other structured API. Here are a few examples:
https://github.com/aspnet/Hosting/blob/c9a4ee8b719ffb01eb44705d68efd158eb80c221/src/Microsoft.AspNetCore.TestHost/ResponseStream.cs
https://github.com/aspnet/HttpSysServer/blob/9587c58140c4f52e616fc89b1c2ada50ed6f0e45/src/Microsoft.AspNetCore.Server.HttpSys/RequestProcessing/ResponseBody.cs

These streams may choose to provide some buffering like in Kestrel as a performance optimization, but they do not need to buffer the whole response. Instead they put back pressure on the caller when they fall behind (e.g. WriteAsync's Task does not complete).

Any solution that requires awareness in the higher level frameworks like MVC will be much more difficult to adopt.

**Comment by Danielku15** [comment]
> The simplest and most compatible would be for your custom IServer to implement a custom Stream rather than using MemoryStream. 
It is exactly my goal to make a custom IServer implementation. But the system underneath does not give me access to a raw stream where I could write the data into. To be precise: I am receiving and sending my "HTTP requests/responses" via a WCF Channel. WCF is used as via duplex channel as reverse proxy, so actually the service accepting my requests is in DMZ machine and my API is running in the LAN. WCF is a passive system: I need to pass a stream to WCF in my models so that WCF can read it and forward it via the streaming channel. 

It sounds interesting that the stream can somehow block the caller. But I fear that will not work in the whole request pipeline. When implementing an IServer, you will call ProcessRequest which will need to complete properly to have all header data available. E.g. if you have a ActionFilter on your API it might not be executed yet when you block the further writing. The `IActionResultExecutor` needs to complete so that the request is known to be completed on the IServer implementation level.

**Comment by Tratcher** [comment]
> But I fear that will not work in the whole request pipeline. 

I don't follow, that's how the existing servers behave today. ProcessRequest does not need to complete before the response can be started, the response is started by the first write to the response body. All response headers must be set prior to the first write. ProcessRequest is only expected to complete when the response body is complete (or there is no body).

For reference, this behavior has been the same for Microsoft.Owin, System.Web, and Microsoft.AspNetCore. Only System.Net.Http (WebApi) ever used the 'double-tap' system where the app was called back to produce the body. That model has some advantages, but it does not compose well with the other systems and would require a major framework overhaul to support.

**Comment by Danielku15** [comment]
> ProcessRequest does not need to complete before the response can be started, the response is started by the first write to the response body. 
The problem is that in this scenario the request needs to complete otherwise you cannot pass the final objects back to WCF with only the stream left to complete. Yes, theoretically it is possible to decouple the whole chain that in the end the stream reading of WCF will allow the OWIN pipeline to block and continue on demand. But synchronizing this stuff will be a nightmare (debugging and readability wise). Maybe I am also to unexperienced in the task-async and continuation world but I see it really as complex. 

On high level that would mean: First you would need to create an applciation context and initiate the request execution without awaiting it. Then you need to await on the stream that the body starts and block the OWIN pipeline from further writing. The IServer layer then needs to continue and give back the "incomplete" object to WCF. You cannot do this earlier because WCF would start serializing the header values before they are available. Awaiting of the first body data is needed. 
WCF would then do the reading on the body stream which allows the OWIN pipeline to continue. Once WCF finishes (dispose on the stream), you need to jump back to the IServer to do the remaining cleanup on the application context. As said: possible but seems like a nightmare to me.

For the most existing IServer implementations this is not an issue as the raw underlying stream can be passed to OWIN. It will do the correct blocking and you can cleanly work with awaits on higher levels.

**Comment by Tratcher** [comment]
We've implemented at least 6 servers using almost the same pattern so far. While it is not trivial, it is certainly manageable. It also isolates a lot of the complexity from the higher level components and application.  Give it a try and we can discuss the details.

**Comment by mkArtakMSFT** [comment]
Closing this as there is no actions we intend to take here.

**Comment by Danielku15** [comment]
It's quite unfortunate that during the active discussion the issue is simply closed. 

I had a look at the following 5 server implementations and all beside the TestHost were based on the fact that there must be a way to actively write the response. Blocking happens on this active writing basis and not on passive reading. 

- Kestrel - Based on raw streams and sockets. (SocketConnection, TransportConnection)
- HttpSysServer - Based on active writes via HttpSendResponseEntityBody. 
- IISIntegration - Based on active writes via AsyncIOEngine.
- NoWin - Based on raw streams and sockets
- TestHost - Closest to a passive system via pipes.

There is no server so far which uses a potential other messaging system than raw writing to a pipe/stream. There are many messaging systems in the market (WCF, MQTT, Azure Message Bus,..) and none of them would provide you a raw stream to write into. You need to use the messaging system hence they simply cannot be used together with the new ASP.net MVC/WebApi. 

Also in the old ASP.net WebApi with the passive stream approach, most responses were anyhow buffered in-memory. Only file and stream results needed to be piped to the raw output. That's why I asked for a simple extension on IFeature basis which would establish a similar system for this streaming use-case.  Unfortuantely the discussion on this defect went a bit too far into the active/passive discussion and whether it is needed and how it can be achieved. 

Likely I will patch the related "ActionExecutors" with this new IFeature I proposed and utilize it in my project.

Thanks for the discussion.

**Comment by Tratcher** [comment]
How is your scenario different from HttpSendResponseEntityBody? E.g. WriteAsync triggers a call to SendMessage.
